### PR TITLE
amber mbar parser

### LINKED ISSUE #42: AMBER FEP parser
In line with the overall [API proposal](http://alchemlyb.readthedocs.io/en/latest/api_proposal.html), we want to have parsers for each of the major MD engines, and eventually have coverage for all of those in use. For Amber we already have a TI parser (#10), we still need an "MBAR/FEP" parser:

- [ ] Extracting reduced potentials `u_nk` from output files (for FEP).

This issue is the nexus for discussion for such parsers for the **AMBER** package. If you have existing parsing code for this package, comment below and we can begin adapting it into the parsers outlined above in a PR.

### PR DESCRIPTION
**Pull request by brycestx**
Fixes #42 

This pull request adds support for amber mbar parsing for amber .out  files by using the [extract_u_nk](https://github.com/alchemistry/alchemlyb/blob/amber-mbar-parse/src/alchemlyb/parsing/amber.py#L250) function, which I wrote to primarily mimic the output format of the [extract_u_nk](https://github.com/alchemistry/alchemlyb/blob/amber-mbar-parse/src/alchemlyb/parsing/gmx.py#L15) gromacs function. 

A [test_u_nk](https://github.com/alchemistry/alchemlyb/blob/amber-mbar-parse/src/alchemlyb/tests/parsing/test_amber.py#L35) function was also added to [test_amber.py](https://github.com/alchemistry/alchemlyb/blob/amber-mbar-parse/src/alchemlyb/tests/parsing/test_amber.py). I am now working on adding a minimal test example to [alchemtest](https://github.com/alchemistry/alchemtest).

### CODE REVIEW

**Comment by orbeckst** [comment]
I'll review once you have added the data files to alchemtest alchemistry/alchemtest#20

**Comment by brycestx** [comment]
@orbeckst done

**Comment by orbeckst** [comment]
Thanks, I reviewed alchemistry/alchemtest#21 – if you can quickly address my comments then we can merge it in a few minutes.

**Comment by brycestx** [comment]
@orbeckst just updated my [alchemistry PR](https://github.com/alchemistry/alchemtest/pull/21)

**Comment by orbeckst** [comment]
alchemistry/alchemtest#21 was merged. I will restart the tests so that the new data files are picked up and then we'll see.

File: src/alchemlyb/parsing/amber.py [hunk 1]
```diff
@@ -199,13 +239,52 @@ def file_validation(outfile):
...
+
+                E_ref = mbar[file_datum.mbar_lambda_idx]
+
+                for lmbda, E in enumerate(mbar):
+                    if E > 0.0:
+                        high_E_cnt += 1
+
+                    file_datum.mbar_energies[lmbda].append(E - E_ref)
+
+        if high_E_cnt:
```
Comment (orbeckst) [inline src/alchemlyb/parsing/amber.py]: add test file or remove from coverage with `#pragma: no cover`

File: src/alchemlyb/parsing/amber.py [hunk 2]
```diff
@@ -199,13 +239,52 @@ def file_validation(outfile):
...
+    file_datum.have_mbar = have_mbar
     return file_datum
 
+
+def extract_u_nk(outfile):
+    file_datum = file_validation(outfile)
+    if not file_validation(outfile):
+        return None
+    if not file_datum.have_mbar:
+        logging.error('ERROR: No MBAR energies found! Cannot parse file.')
```
Comment (orbeckst) [inline src/alchemlyb/parsing/amber.py]: SHouldn't this raise an exception? At this point it shouldn't continue, should it?

Comment (orbeckst) [reply src/alchemlyb/parsing/amber.py]: It should behave in the same way as the TI parser.

**Comment by brycestx** [comment]
@orbeckst [alchemistry/alchemtest#22](https://github.com/alchemistry/alchemtest/pull/22) must be approved before the additional tests I have added can be tested properly

**Comment by orbeckst** [comment]
I merged https://github.com/alchemistry/alchemtest/pull/22 and restarted the Travis tests. Will have a look once the tests are done.

File: src/alchemlyb/parsing/amber.py [hunk 3]
```diff
@@ -189,6 +203,32 @@ def file_validation(outfile):
...
+                                                      ['ifmbar',
+                                                        'bar_intervall'],
+                                                      '^---')
+        if have_mbar:
+            mbar_ndata = int(nstlim / mbar_ndata)
+            mbar_lambdas = _process_mbar_lambdas(secp)
+            file_datum.mbar_lambdas = mbar_lambdas
+            clambda_str = '%6.4f' % clambda
+
+            if clambda_str not in mbar_lambdas:
```
Comment (orbeckst) [inline src/alchemlyb/parsing/amber.py]: This block is not tested. See https://codecov.io/gh/alchemistry/alchemlyb/pull/51/src/src/alchemlyb/parsing/amber.py?before=src/alchemlyb/parsing/amber.py#L218

Put a `#pragma: no cover` to indicate that we exclude it from test.

Comment (brycestx) [reply src/alchemlyb/parsing/amber.py]: Should be addressed in [alchemistry/alchemtest#23](https://github.com/alchemistry/alchemtest/pull/23)

**Comment by brycestx** [comment]
@orbeckst should be good to go after [alchemistry/alchemtest#23](https://github.com/alchemistry/alchemtest/pull/23) merge

**Comment by orbeckst** [comment]
Merged PR https://github.com/alchemistry/alchemtest/pull/24 and waiting for Travis.

**Comment by orbeckst** [comment]
Travis has issues https://www.traviscistatus.com/incidents/c1s4dlyxd4lf so might take a while.

**Comment by brycestx** [comment]
Yea I saw it hanging and was hoping it wasn't my code that caused it... ha

**Comment by orbeckst** [comment]
Thanks @brycestx ; I will take it from here.

I will rebase and then merge. (Probably also add another `no cover`.)

**Comment by orbeckst** [comment]
I rebased against master and excluded the `file_validation(outfile)` in `extract_dHdl()` from the coverage. (Note, I force-pushed to the branch!)

I'll wait for the tests to finish and then squash & merge.

**Comment by orbeckst** [comment]
I squashed everything into a single commit.

**Comment by orbeckst** [comment]
@brycestx merged – thanks for all the work.

(The coverage diff was "only" 91% but I looked at the code and it seemed ok to me so I overrode the merge warning.)

**Comment by brycestx** [comment]
@orbeckst hmm any idea why that is. Looking here it looks much better than before.. https://codecov.io/gh/alchemistry/alchemlyb/src/ab1b6573670ab13d09b92ee291e9159a1e3ca860/src/alchemlyb/parsing/amber.py

**Comment by orbeckst** [comment]
The relative diff is also 91.66% in the page that you linked – I think we've been looking at the same thing. The overall coverage of the file is still 97.35%. If you want to test the lines that are omitted or marked `# pragma: no cover` then that would be great, of course, but I consider the code that went in of a high standard with good testing already.

**Pull request merged**